File: Array&String/HashTable/Python/sort_array_by_increasing_frequency.py
from collections import Counter
from typing import List
# 1636. Sort Array by Increasing Frequency
# https://leetcode.com/problems/sort-array-by-increasing-frequency/
# Given an array of integers nums, sort the array in increasing order based on the frequency of the values.

class Solution:
    def frequencySort(self, nums: List[int]) -> List[int]:
        freq = Counter(nums)
        freqToNums = {}
        for num, freq in freq.items():
            freqToNums[freq] = freqToNums.get(freq, []) + [num]

        res = []
        for f in sorted(freqToNums.keys()):
            numsAtFreq = sorted(freqToNums[f], reverse=True)
            for num in numsAtFreq: res.extend([num] * f)
        return res

    # using sort + lambda function
    def frequencySort2(self, nums: List[int]) -> List[int]:
        freq = Counter(nums)
        return sorted(nums, key=lambda x: (freq[x], -x))

File: Array&String/HashTable/CPP/sort_array_by_increasing_frequency.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <functional>
using namespace std;
// 1636. Sort Array by Increasing Frequency
// https://leetcode.com/problems/sort-array-by-increasing-frequency/description/
// Given an array of integers nums, sort the array in increasing order based on the frequency of the values.

class Solution {
    public:
        vector<int> frequencySort(vector<int>& nums) {
            unordered_map<int, int> freqs;
            for (const int x : nums) freqs[x]++;

            unordered_map<int, vector<int>> freqToNums;
            for (const auto& [n, freq] : freqs) {
                freqToNums[freq].push_back(n);
            }

            vector<int> res;
            for (auto& [f, numsAtFreq] : freqToNums) {
                sort(numsAtFreq.begin(), numsAtFreq.end(), greater<>());
                for (const int num : numsAtFreq) {
                    res.insert(res.end(), f, num);
                }
            }

            return res;
        }
    };

    // using sort with custom comparator
    class Solution2 {
        public:
            vector<int> frequencySort(vector<int>& nums) {
                unordered_map<int, int> freqs;
                for (const int x : nums) freqs[x]++;

                sort(nums.begin(), nums.end(), [&](int a, int b) {
                    return freqs[a] == freqs[b] ? a > b : freqs[a] < freqs[b];
                });

                return nums;
            }
        };
